File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque

# 올리기 & 타기 1초, length 당 1초
def solution(bridge_length, weight, truck_weights):
    truck_weights = deque(truck_weights)  # 기존 트럭 deque로 바꾸기
    n = len(truck_weights)  # 총 트럭 개수

    current_time = 1  # 총 시간

    truck_queue = deque()  # 다리 위 트럭
    time_queue = deque()  # 다리 위 트럭 시간

    while truck_weights or truck_queue:
        # 트럭이 남아있고
        # 예상 길이 이하인 경우 & 예상 무게 이하
        if truck_weights and len(truck_queue) <= bridge_length - 1 and sum(truck_queue) + truck_weights[0] <= weight:
            truck_queue.append(truck_weights.popleft())
            time_queue.append(1)
        # 미리 간 차가 다리를 다 건넜는지 확인
        m = len(time_queue)
        comp = False
        for i in range(m):
            if time_queue[i] == bridge_length:
                comp = True
            else:
                time_queue[i] += 1
        # 다 건넜으면 하나 빼기
        if comp:
            truck_queue.popleft()
            time_queue.popleft()
        # 시간 증가
        current_time += 1
    return current_time

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.js
function solution(bridge_length, weight, truck_weights) {
  // 다리 길이만큼 0으로 채운 배열 만들기
  let bridge = Array(bridge_length).fill(0)
  let current_weight = 0;  // 현재 다리 위 무게
  let current_time = 0;  // 현재 시간
  // 대기 중이 트럭이 있거나 현재 다리를 건너는 트럭이 있으면
  while (truck_weights.length > 0 || current_weight > 0) {
    // 시간 증가
    current_time += 1
    // 다리에 있는 트럭 앞으로 보내고 그만큼 무게 제거
    current_weight -= bridge.shift()
    // 대기 트럭이 있고 대기 트럭 가장 첫 트럭 무게가 감당 무게 안이면
    if (truck_weights) {
      if (truck_weights[0] + current_weight <= weight) {
        let truck = truck_weights.shift()
        bridge.push(truck)
        current_weight += truck
      } else {
      bridge.push(0)
      }
    }
  }
  return current_time;
}